learner.py

```diff
@@ -1,24 +1,38 @@
 import numpy as np
 from lib import *
 
+cost_power = 10
+
 def cost(Y, Yp):
-    return np.sum(np.square(Y-Yp))/2
+    return np.sum(np.power(Y-Yp, cost_power))/cost_power
+
+def cost_gradient(Y, Yp):
+    return np.power(Yp - Y, cost_power-1)
 
-def gradient_step(mp, X, Y, lrate):
+def gradient(mp, X, Y):
     sc = len(X)
     Yp, grad = mp.gradient(X)
-    step = np.sum((Yp - Y)*grad, axis=1)
+    step = np.sum(cost_gradient(Y, Yp)*grad, axis=1)
     step /= np.linalg.norm(step)
-    step *= -lrate
-    nmp = mp.new_with(mp.params + step)
-    nmp.cost = cost(Y, Yp)
-    return nmp
+    return step, Yp
 
-def learn(mp, X, Y, lrate=0.3, rep=1000):
+def learn(mp, X, Y, lrate=0.3, rep=1000, beta1=0.9, beta2=0.999, eps=1e-4):
     X = mp.normalize_input(X)
     Y = np.array(Y, dtype=np.float64)
     best = mp
+    m = np.zeros_like(mp.params)
+    v = np.zeros_like(mp.params)
     for repi in range(rep):
-        mp = gradient_step(mp, X, Y, lrate)
+        step, Yp = gradient(mp, X, Y)
+        
+        # Adam optimizer
+        m = beta1 * m + (1 - beta1) * step
+        v = beta2 * v + (1 - beta2) * np.power(step, 2)
+        m_hat = m / (1 - np.power(beta1, repi+1))
+        v_hat = v / (1 - np.power(beta2, repi+1))
+        step = m_hat / (np.sqrt(v_hat) + eps)
+        
+        mp = mp.new_with(mp.params - lrate * step)
+        mp.cost = cost(Y, Yp)
```

Comment:
`Yp` is still from the old model here, so cost is not accurate.
Issue existed before but it seems like you tried to fix it. I avoided a second prediction run because that would essentially double runtime for no real gain (except that `best` would actually do something useful since without ADAM a common oscillation period is 2).

Reply:
Fixed